### PR TITLE
Add Gerrit chart to incubator

### PR DESCRIPTION
**Pull request by wbrefvem**
#### What this PR does / why we need it:

Adds a Gerrit Code Review chart.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Variables are documented in the README.md

### CODE REVIEW

File: incubator/gerrit/Chart.yaml
Comment (giacope) [inline incubator/gerrit/Chart.yaml]: Since this is a new chart, the version should start at `0.1.0`.

File: incubator/gerrit/Chart.yaml
Comment (giacope) [inline incubator/gerrit/Chart.yaml]: Shouldn't this be `2.15.3`

**Review by giacope — CHANGES_REQUESTED** [review]

File: incubator/gerrit/Chart.yaml
Comment (wbrefvem) [reply incubator/gerrit/Chart.yaml]: Indeed it should. Thanks!

File: incubator/gerrit/Chart.yaml
Comment (wbrefvem) [reply incubator/gerrit/Chart.yaml]: Noted, I'll make the change.

File: incubator/gerrit/templates/deployment.yaml
Comment (hemantsonawane95) [inline incubator/gerrit/templates/deployment.yaml]: Please fix typo

File: incubator/gerrit/templates/deployment.yaml
Comment (unguiculus) [inline incubator/gerrit/templates/deployment.yaml:29]: I general, we suggest not to set defaults for limits. You may add suggestions to the values.yaml an comment them out. That being said, 1Gi is quite a lot for setting permissions.

Have you thought about using a security context?

File: incubator/gerrit/templates/deployment.yaml
Comment (unguiculus) [inline incubator/gerrit/templates/deployment.yaml]: Use `apps/v1`

File: incubator/gerrit/templates/pvc.yaml
Comment (unguiculus) [inline incubator/gerrit/templates/pvc.yaml:9]: You explicitly want multiple volumes instead of having everything on one volume? Just asking...

File: incubator/gerrit/templates/pvc.yaml
Comment (unguiculus) [inline incubator/gerrit/templates/pvc.yaml:20]: You allow multiple replicas. This certainly won't work with `ReadWriteOnce`. Does Gerrit support running with multiple replicas anyways? Wouldn't you need some kind of replication setup instead?

File: incubator/gerrit/values.yaml
Comment (unguiculus) [inline incubator/gerrit/values.yaml]: Don't specify defaults for resources. Comment them out.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please add a readme file and move the chart to stable directly.

Helm was updated to match the new Kubernetes labels convention. If you use `helm create` with Helm 2.11 you'll get a new chart with these new labels. Would you mind updating the chart accordingly?

File: incubator/gerrit/templates/deployment.yaml
Comment (wbrefvem) [reply incubator/gerrit/templates/deployment.yaml:29]: I had not looked at a security context, will definitely take a look.

As for resource limits, if they are configurable and don't have defaults, won't `helm install .` fail?

File: incubator/gerrit/templates/deployment.yaml
Comment (wbrefvem) [reply incubator/gerrit/templates/deployment.yaml]: Noted, will make the change.

File: incubator/gerrit/templates/pvc.yaml
Comment (wbrefvem) [reply incubator/gerrit/templates/pvc.yaml:9]: Yeah, this follows the architecture that the Gerrit community recommends for a vanilla Docker setup (https://hub.docker.com/r/gerritcodereview/gerrit), where each of 4 directories under the Gerrit home directory has its own volume. I wasn't sure whether this was optimal for Kubernetes, and in fact it makes initializing Gerrit more complicated, but I figured it was a good starting point. I'm open to suggestions if you have any.

File: incubator/gerrit/templates/pvc.yaml
Comment (wbrefvem) [reply incubator/gerrit/templates/pvc.yaml:20]: That's an excellent point. It was not a conscious choice to support multiple replicas. IIRC this was a default from `helm create` that I never changed. I think for now, supporting a single replica is sufficient, and for future versions I may try to tackle multiple replicas.

File: incubator/gerrit/templates/deployment.yaml
Comment (wbrefvem) [reply incubator/gerrit/templates/deployment.yaml:29]: Okay, nvm about the resources. I see how `helm create` does it, so I'm reverting to that.

**Comment by wbrefvem** [comment]
@unguiculus I've made the changes you requested. Please let me know if you have any further feedback.

File: stable/gerrit/templates/deployment.yaml
Comment (hemantsonawane95) [inline stable/gerrit/templates/deployment.yaml]: It should be /var/gerrit/{{ .name }} if you want to be consistent

Comment (wbrefvem) [reply stable/gerrit/templates/deployment.yaml]: Ah, thanks. Good catch.

File: stable/gerrit/templates/ingress.yaml
Comment (unguiculus) [inline stable/gerrit/templates/ingress.yaml]: Make these annotations configurable. Don't specify them as default. Folks may want to use ingress controller other than Nginx.

File: stable/gerrit/templates/service.yaml
Comment (unguiculus) [inline stable/gerrit/templates/service.yaml]: It doesn't make sense to make target ports configurable since the are hard-coded anyways in the deployment. Just reference the ports via their names here.

File: stable/gerrit/templates/deployment.yaml
Comment (unguiculus) [inline stable/gerrit/templates/deployment.yaml]: Wrong labels in selector. Should be:

```yaml
app.kubernetes.io/name: {{ include "gerrit.name" . }}
app.kubernetes.io/instance: {{ .Release.Name }}
```

File: stable/gerrit/templates/service.yaml
Comment (unguiculus) [inline stable/gerrit/templates/service.yaml]: Wrong labels in selector. Should be:

```yaml
app.kubernetes.io/name: {{ include "gerrit.name" . }}
app.kubernetes.io/instance: {{ .Release.Name }}
```

File: stable/gerrit/values.yaml
Comment (unguiculus) [inline stable/gerrit/values.yaml]: Don't specify a default password. Leave it empty.

File: stable/gerrit/templates/deployment.yaml
Comment (unguiculus) [inline stable/gerrit/templates/deployment.yaml]: Please use consistent formatting: Space after `{{` and before `}}`.

File: stable/gerrit/templates/deployment.yaml
Comment (unguiculus) [inline stable/gerrit/templates/deployment.yaml]: You should be able to remove the init container if you add this instead:

```yaml
securityContext:
  runAsUser: 1000
  fsGroup: 1000
  runAsNonRoot: true
```

**Review by unguiculus — CHANGES_REQUESTED** [review]
```
Ingress.extensions "gerrit" is invalid: spec: Invalid value: []extensions.IngressRule(nil): either `backend` or `rules` must be specified
```
Fix the ingress, make the ingress optional and disabled by default.

Please add new changes on top. Don't rebase the branch. Otherwise GitHub doesn't show new changes. In the end we squash on merge anyways.

File: stable/gerrit/templates/deployment.yaml
Comment (unguiculus) [inline stable/gerrit/templates/deployment.yaml]: Wrong labels. Should be:

```yaml
app.kubernetes.io/name: {{ include "gerrit.name" . }}
app.kubernetes.io/instance: {{ .Release.Name }}
```

**Comment by wbrefvem** [comment]
@unguiculus Would it be possible to merge this chart under incubator and make these changes later, like before moving it to stable? My goal for this PR is not to have a finished, stable chart, but just a minimally usable one as a starting point, which is why I initially put it incubator.

**Comment by unguiculus** [comment]
We tend to not add more charts to incubator. The charts repo should be a place for half-baked stuff. I'd like you to consider my comments before we merge.

**Comment by wbrefvem** [comment]
@unguiculus Will you approve and/or merge this PR if I make the changes you request? I'm not going to waste my time if not.

**Comment by unguiculus** [comment]
Now I wonder why you moved it back to incubator. I suggested the opposite.

Regarding your "wasting time." With all due respect this sounds like you just want to drop this here and leave it alone. If you add a chart you're the maintainer. You're expected to review PRs and triage issues for it. Are you willing to do that? I'd really appreciate having a Gerrit chart but I don't want you to drop unfinished stuff and then walk away.

**Comment by wbrefvem** [comment]
@unguiculus I understand your concern, and I appreciate your feedback. I have no intention of abandoning an unfinished chart. My intent in creating it is to use it for a Gerrit addon in Jenkins X, and it could also be useful in improving Gerrit support for Prow.

The reason I want to put it in incubator is so that the community is aware that it's new and has a chance to provide feedback. I don't want to mark it as stable until we've had a chance to get some feedback from users. My hope is that it would very quickly move into stable.

**Pull request closed without merging**
